File: 백준/Gold/1915. 가장 큰 정사각형/가장 큰 정사각형.py
n, m = map(int,input().strip().split())

grid = []
grid.append(['0'] * (m + 1))
for _ in range(n):
  grid.append(list('0' + input().strip()))

dp = [[0] * (m + 1) for _ in range(n + 1)]

max_len = 0
for i in range(1,n + 1):
  for j in range(1,m + 1):
    if grid[i][j] == '1':
      dp[i][j] = min(
        dp[i-1][j],     # 위
        dp[i][j-1],     # 왼쪽
        dp[i-1][j-1]    # 왼쪽 위
        ) + 1
      max_len = max(max_len, dp[i][j])

print(max_len ** 2)

File: 백준/Gold/1915. 가장 큰 정사각형/가장 큰 정사각형.js
const input = require('fs')
	.readFileSync('/dev/stdin', 'utf-8')
	.trim()
	.split('\n');

const [n, m] = input[0].trim().split(' ').map(Number);
const grid = [];
grid.push(Array(m + 1).fill('0'));
for (let i = 1; i < n + 1; i++) {
	grid.push(['0', ...input[i].trim().split('')]);
}

const dp = Array.from({ length: n + 1 }, () => Array(m + 1).fill(0));

let maxLen = 0;
for (let i = 1; i < n + 1; i++) {
	for (let j = 1; j < m + 1; j++) {
		if (grid[i][j] === '1') {
			dp[i][j] = Math.min(dp[i - 1][j], dp[i - 1][j - 1], dp[i][j - 1]) + 1;
			maxLen = Math.max(maxLen, dp[i][j]);
		}
	}
}

console.log(Math.pow(maxLen, 2));
